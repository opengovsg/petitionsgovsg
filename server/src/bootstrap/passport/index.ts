import express from 'express'
import passport from 'passport'
import { sgidStrategy } from './sgid.strategy'
import { AuthUserDto } from '~shared/types/api'

const privateKeyPem = (process.env.SGID_PRIV_KEY ?? '').replace(/\\n/g, '\n')

export const passportConfig = (app: express.Application): void => {
  sgidStrategy(privateKeyPem)
  app.use(passport.initialize())
  app.use(passport.session())

  // stores user to session
  passport.serializeUser((user, done) => {
    done(null, user)
  })

  // retrieves user from session
  passport.deserializeUser((user: AuthUserDto, done) => {
    done(null, user)
  })
}
