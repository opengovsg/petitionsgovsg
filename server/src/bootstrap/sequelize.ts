import { Sequelize } from 'sequelize'

import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { dbConfig } from './config/database'

export const sequelize = new Sequelize({ ...dbConfig, logging: false })

export const { Addressee } = defineAddressee(sequelize)
export const { Post } = definePost(sequelize, { Addressee })
export const Signature = defineSignature(sequelize, { Post })
export const Subscription = defineSubscription(sequelize, { Post })

export default sequelize
