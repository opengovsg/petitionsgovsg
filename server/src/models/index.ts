export { Signature, defineSignature } from './signatures.model'
export { Post, definePost } from './posts.model'
