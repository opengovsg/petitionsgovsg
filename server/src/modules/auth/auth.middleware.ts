import { StatusCodes } from 'http-status-codes'
import { ControllerHandler } from '../../types/response-handler'
import { decodeUserJWT } from '../../util/jwt'
export class AuthMiddleware {
  /**
   * Middleware that only allows authenticated users to pass through to the next
   * handler.
   * @returns next if user exists in session
   * @returns 401 if user does not exist in session
   */
  authenticate: ControllerHandler = (req, res, next) => {
    // Check if user is authenticated
    try {
      const token = req.cookies.jwt
      req.user = decodeUserJWT(req)
    } catch (error) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in.' })
    }
    return next()
  }
}
