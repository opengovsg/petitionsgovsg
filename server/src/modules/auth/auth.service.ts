import { Post, PostStatus } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'

export class AuthService {
  private Post: ModelDef<Post>

  constructor({ Post }: { Post: ModelDef<Post> }) {
    this.Post = Post
  }

  /**
   * Check if a user has permission to edit a post
   * @param userId of user
   * @param postId of post
   * @returns true if user has permission to answer post
   */
  hasPermissionToEditPost = async (
    userId: number,
    postId: number,
  ): Promise<boolean> => {
    // TODO:
    return true
  }

  /**
   * Check if a user is able to view the post
   * @param post post to be viewed
   * @param userId id of user
   * @returns true if user can view post
   */
  verifyUserCanViewPost = async (
    post: Post,
    userId?: string,
  ): Promise<void> => {
    // If post is public, anyone can view
    if (post.status === PostStatus.Open) return

    // If private or archived, must be logged in
    if (!userId)
      throw new Error('User must be logged in to access private post')

    // TODO
    // const user = await this.User.findOne({ where: { id: userId } })

    // if (user) {
    //   // If officer, they may have permission to answer
    //   if (await this.hasPermissionToEditPost(user.id, post.id)) return

    //   // If none of the above, they must have created the post
    //   if (user.id === post.userId) return
    //   throw new Error('User does not have permission to access post')

    return
  }
}
