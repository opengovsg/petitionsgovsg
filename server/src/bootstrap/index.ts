import compression from 'compression'
import cors from 'cors'
import express from 'express'
import fs from 'fs'
import helmet from 'helmet'
import { StatusCodes } from 'http-status-codes'
import path from 'path'
import { checkOwnershipUsing } from '../middleware/checkOwnership'
import { SignatureController } from '../modules/signatures/signature.controller'
import { SignatureService } from '../modules/signatures/signature.service'
import { AuthController } from '../modules/auth/auth.controller'
import { AuthMiddleware } from '../modules/auth/auth.middleware'
import { AuthService } from '../modules/auth/auth.service'
import { PostController } from '../modules/post/post.controller'
import { PostService } from '../modules/post/post.service'
import { UserService } from '../modules/user/user.service'
import { api } from '../routes'
import { baseConfig, Environment } from './config/base'
import { emailValidator } from './email-validator'
import { helmetOptions } from './helmet-options'
import { requestLoggingMiddleware } from './logging/request-logging'
import { passportConfig } from './passport'
import { Signature, Post, sequelize, User } from './sequelize'
import sessionMiddleware from './session'

export { sequelize } from './sequelize'
export const app = express()

// compressing api response
app.use(compression())

// logger
app.use(requestLoggingMiddleware)

// cors enable
app.options('*', cors<express.Request>())
app.use(cors({ origin: `http://localhost:${process.env.SERVER_PORT}` }))

// security config
app.use(helmet(helmetOptions))

// body-parser
app.use(express.urlencoded({ extended: true }))
app.use(express.json())

// passport and session
app.set('trust proxy', 1) // trust first proxy
app.use(sessionMiddleware(sequelize))
passportConfig(app, User)

// all the api routes
const authService = new AuthService({
  emailValidator,
  User,
  Post,
})
const authMiddleware = new AuthMiddleware()
const postService = new PostService({
  Signature,
  Post,
  User,
  sequelize,
})
const signatureService = new SignatureService({
  Post,
  Signature,
  sequelize,
})

const apiOptions = {
  signature: {
    controller: new SignatureController({
      signatureService,
      postService,
    }),
    authMiddleware,
    checkOwnership: checkOwnershipUsing({ Post, Signature, User }),
  },
  auth: {
    controller: new AuthController(),
    authMiddleware,
  },
  post: {
    controller: new PostController({
      authService,
      postService,
    }),
    authMiddleware,
  },
}

app.use('/api/v1', api(apiOptions))

// connection with client setup
if (baseConfig.nodeEnv === Environment.Prod) {
  const setNoCache = (res: express.Response) => {
    const date = new Date()
    date.setFullYear(date.getFullYear() - 1)
    res.setHeader('Expires', date.toUTCString())
    res.setHeader('Pragma', 'no-cache')
    res.setHeader('Cache-Control', 'public, no-cache')
  }

  const setLongTermCache = (res: express.Response) => {
    const date = new Date()
    date.setFullYear(date.getFullYear() + 1)
    res.setHeader('Expires', date.toUTCString())
    res.setHeader('Cache-Control', 'public, max-age=31536000, immutable')
  }

  app.use(
    express.static(path.resolve(__dirname, '../../..', 'client', 'build'), {
      extensions: ['html'],
      setHeaders(res, path) {
        if (path.includes('static')) {
          setLongTermCache(res)
        } else {
          setNoCache(res)
        }
      },
    }),
  )

  const index = fs.readFileSync(
    path.resolve(__dirname, '../../..', 'client', 'build', 'index.html'),
  )

  // TODO: sync-up with front end
  // iterate through list of possible paths to serve index file
  const allStaticPaths = [
    '/',
    '/questions',
    '/login',
    '/add/question',
    '/edit/question/:id',
    '/terms',
    '/agency-terms',
    '/privacy',
    '/agency-privacy',
    '/user-login',
    '/unauthorised',
  ]
  for (const path of allStaticPaths) {
    app.get(path, (_req, res) =>
      res.header('content-type', 'text/html').send(index),
    )
  }

  app.get('/not-found', (_req, res) =>
    res
      .header('content-type', 'text/html')
      .status(StatusCodes.NOT_FOUND)
      .send(index),
  )

  app.get('*', (_req, res) =>
    res
      .header('content-type', 'text/html')
      .redirect(StatusCodes.MOVED_PERMANENTLY, '/not-found'),
  )
}

export default app
