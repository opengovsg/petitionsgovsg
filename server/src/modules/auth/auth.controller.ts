import { StatusCodes } from 'http-status-codes'
import { isEmpty } from 'lodash'
import passport from 'passport'
import {
  callbackRedirectUnauthorisedURL,
  callbackRedirectURL,
} from '../../bootstrap/config/auth'
import { Message } from 'src/types/message-type'
import { ErrorDto, LoadPublicUserDto, UserAuthType } from '~shared/types/api'
import { createLogger } from '../../bootstrap/logging'
import { UserService } from '../../modules/user/user.service'
import { ControllerHandler } from '../../types/response-handler'

const logger = createLogger(module)

export class AuthController {
  private userService: Public<UserService>

  constructor({ userService }: { userService: Public<UserService> }) {
    this.userService = userService
  }

  /**
   * Fetch logged in user details after being authenticated.
   * @returns 200 with user details
   * @returns 500 if user id not found
   * @returns 500 if database error
   */
  loadUser: ControllerHandler<unknown, LoadPublicUserDto | ErrorDto> = async (
    req,
    res,
  ) => {
    const id = req.user?.id
    const type = req.user?.type
    if (!id) {
      logger.error({
        message: 'User not found after being authenticated',
        meta: {
          function: 'loadUser',
          userId: req.user?.id,
        },
      })
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }

    if (type === UserAuthType.Public) {
      try {
        const user = await this.userService.loadUser(id)
        if (user) {
          return res.status(StatusCodes.OK).json({ id: user.id })
        } else {
          logger.error({
            message: 'User not found',
            meta: {
              function: 'loadUser',
              userId: req.user?.id,
            },
          })
          return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
        }
      } catch (error) {
        logger.error({
          message: 'Database Error while loading public user',
          meta: {
            function: 'loadUser',
            userId: req.user?.id,
          },
          error,
        })
        return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
      }
    }

    if (type === UserAuthType.Public) {
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(null)
    }
  }

  /**
   * Logout
   * @returns 200 if logged out
   */
  handleLogout: ControllerHandler = (req, res) => {
    if (!req.session || isEmpty(req.session)) {
      logger.error({
        message: 'Attempted to sign out without a session',
        meta: {
          function: 'handleLogout',
        },
      })
      return res.sendStatus(StatusCodes.BAD_REQUEST)
    }

    req.session.destroy((error) => {
      if (error) {
        logger.error({
          message: 'Failed to destroy session',
          meta: {
            action: 'handleLogout',
            function: 'handleLogout',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Sign out failed' })
      }

      // No error.
      res.clearCookie('connect.sid')
      return res.status(StatusCodes.OK).json({ message: 'Sign out successful' })
    })
  }

  handleSgidLogin: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { redirect: string }
  > = async (req, res, next) => {
    const { redirect } = req.query
    // store redirect to post in state
    if (redirect) {
      passport.authenticate('sgid-with-name', { state: redirect })(
        req,
        res,
        next,
      )
    } else {
      passport.authenticate('sgid-with-name')(req, res, next)
    }
  }
  /**
   * Verify otp received by the user
   * @params code
   * @params state
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   */
  handleSgidCallback: ControllerHandler<
    undefined,
    undefined,
    undefined,
    { code: string; state: string | undefined }
  > = async (req, res, next) => {
    const { state } = req.query
    passport.authenticate(
      'sgid-with-name',
      {},
      (error, user, info: Message) => {
        if (error) {
          logger.error({
            message: 'Error while authenticating',
            meta: {
              function: 'handleSgidCallback',
            },
            error,
          })
          return res.redirect(callbackRedirectUnauthorisedURL)
        }
        if (!user) {
          logger.warn({
            message: info.message,
            meta: {
              function: 'handleSgidCallback',
            },
          })
          res.redirect(callbackRedirectUnauthorisedURL)
        }
        req.logIn(user, (error) => {
          if (error) {
            logger.error({
              message: 'Error while logging in',
              meta: {
                function: 'handleSgidCallback',
              },
              error,
            })
            return res.redirect(callbackRedirectUnauthorisedURL)
          }
          //
          /**
           * Regenerate session to mitigate session fixation
           * We regenerate the session upon logging in so an
           * anonymous session cookie cannot be used
           */
          const passportSession = req.session.passport
          req.session.regenerate(function (error) {
            if (error) {
              logger.error({
                message: 'Error while regenerating session',
                meta: {
                  function: 'handleSgidCallback',
                },
                error,
              })
              return res.redirect(callbackRedirectUnauthorisedURL)
            }
            //req.session.passport is now undefined
            req.session.passport = passportSession
            req.session.save(function (error) {
              if (error) {
                logger.error({
                  message: 'Error while saving regenerated session',
                  meta: {
                    function: 'handleSgidCallback',
                  },
                  error,
                })
                return res.redirect(callbackRedirectUnauthorisedURL)
              }
              return res.redirect(callbackRedirectURL(state))
            })
          })
        })
      },
    )(req, res, next)
  }
}
