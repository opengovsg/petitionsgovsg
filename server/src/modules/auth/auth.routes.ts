import express from 'express'
import { AuthController } from './auth.controller'
import { AuthMiddleware } from './auth.middleware'

export const routeAuth = ({
  controller,
  authMiddleware,
}: {
  controller: AuthController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()

  /**
   * Fetch logged in user details
   * @route   GET /api/auth
   * @returns 200 with user details
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.get('/', authMiddleware.authenticate, controller.loadUser)

  /**
   * Logout
   * @route   POST /api/auth/logout
   * @returns 200 if logged out
   * @access  private
   */
  router.post('/logout', controller.handleLogout)

  /**
   * Sgid Login
   * @route   POST /api/auth/sgid/login
   * @returns 302 with sgid oidc link
   * @access  private
   */
  router.get('/sgid/login', controller.handleSgidLogin)

  /**
   * Sgid Callback
   * @route   GET /api/auth/callback
   * @returns 302 to home page if successful login
   * @returns 302 to unauthorised page if error
   * @returns 302 to sgid auth page if no state or code params received
   * @access  private
   */
  router.get('/callback', controller.handleSgidCallback)

  return router
}
