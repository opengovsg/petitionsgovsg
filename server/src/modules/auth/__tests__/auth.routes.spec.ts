import express from 'express'
import session from 'express-session'
import { StatusCodes } from 'http-status-codes'
import { ModelCtor, Sequelize } from 'sequelize'
import supertest, { Session } from 'supertest-session'
import { createAuthedSession, logoutSession } from '../../../../tests/mock-auth'
import { passportConfig } from '../../../bootstrap/passport'
import { User as UserModel } from '../../../models'
import { createTestDatabase, getModel, ModelName } from '../../../util/jest-db'
import { AuthController } from '../auth.controller'
import { AuthMiddleware } from '../auth.middleware'
import { routeAuth } from '../auth.routes'
import { parse } from 'querystring'
import passport from 'passport'
import passportCustom from 'passport-custom'
import { UserAuthType } from '~shared/types/api'

describe('/auth', () => {
  const path = '/auth'
  // Mock as few services as possible for testing of authentication flow
  const userService = {
    loadUser: jest.fn(),
    loadUserBySgid: jest.fn(),
  }

  let authController: AuthController

  const authMiddleware = new AuthMiddleware()

  // Set up supertest
  const app = express()
  app.use(express.json())
  let request: Session

  // Set up sequelize
  let db: Sequelize
  let User: ModelCtor<UserModel>
  let mockUser: UserModel

  beforeAll(async () => {
    db = await createTestDatabase()
    User = getModel<UserModel>(db, ModelName.User)
    mockUser = await User.create({
      sgid: 'u=35',
      displayname: 'LIM YONG XIANG',
      fullname: 'LIM YONG XIANG',
      email: 'limyongxiang@test.gov.sg',
      active: true,
    })
    authController = new AuthController({
      userService,
    })

    // passport and session
    app.use(
      session({
        secret: 'secret',
        resave: false,
        saveUninitialized: false,
        store: new session.MemoryStore(),
      }),
    )
    passportConfig(app, User)
    const CustomStrategy = passportCustom.Strategy

    // Mock passport login
    const authedUser: Express.User | undefined = {
      id: 1,
      type: UserAuthType.Public,
    }
    passport.use(
      'mock',
      new CustomStrategy((req, done) => {
        done(null, authedUser)
      }),
    )
    app.get('/auth/mock', passport.authenticate('mock'))

    // passport before route
    app.use(
      path,
      routeAuth({
        controller: authController,
        authMiddleware,
      }),
    )
  })

  beforeEach(() => {
    request = supertest(app)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('loadUser', () => {
    it('should return 400 when not logged in', async () => {
      // Act
      const result = await request.get(path)
      // Assert
      expect(result.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(result.body).toStrictEqual({ message: 'User is unauthorized.' })
    })

    it('should return 200 with user data when logged in', async () => {
      // Arrange
      userService.loadUser.mockReturnValueOnce(mockUser)
      const session = await createAuthedSession(request)

      // Act
      const result = await session.get(path)

      // Assert
      expect(result.status).toEqual(StatusCodes.OK)
      // Body should be an user object.
      expect(result.body).toMatchObject({
        // Required since that's how the data is sent out from the application.
        fullname: mockUser.fullname,
        id: mockUser.id,
      })
    })

    it('should return 400 if logged out after logging in', async () => {
      // Log in user
      // Arrange
      userService.loadUser.mockReturnValueOnce(mockUser)
      const session = await createAuthedSession(request)

      // Act
      const result = await session.get(path)
      // Assert
      expect(result.status).toEqual(StatusCodes.OK)

      // Attempt to load user after logged out
      // Act
      request = await logoutSession(request)
      const logoutResult = await request.get(path)
      // Assert
      expect(logoutResult.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(logoutResult.body).toStrictEqual({
        message: 'User is unauthorized.',
      })
    })
  })

  describe('/sgid/login', () => {
    it('should return redirect with well-formed sgid url', async () => {
      // Act
      const result = await request.get('/auth/sgid/login')
      const resultLocation = parse(result.header.location.split('?')[1])

      // Assert
      expect(result.status).toEqual(StatusCodes.MOVED_TEMPORARILY)
      expect(resultLocation).toHaveProperty('response_type', 'code')
      expect(resultLocation).toHaveProperty('scope', 'openid myinfo.name')
      expect(resultLocation).toHaveProperty('client_id')
      expect(resultLocation).toHaveProperty('redirect_uri')
      expect(resultLocation).toHaveProperty('state')
    })
  })

  describe('/callback', () => {
    it('should return redirect if no code and state params', async () => {
      // Act
      const result = await request.get('/auth/callback')
      const resultLocation = parse(result.header.location.split('?')[1])

      // Assert
      expect(result.status).toEqual(StatusCodes.MOVED_TEMPORARILY)
      expect(resultLocation).toHaveProperty('response_type', 'code')
      expect(resultLocation).toHaveProperty('scope', 'openid myinfo.name')
      expect(resultLocation).toHaveProperty('client_id')
      expect(resultLocation).toHaveProperty('redirect_uri')
      expect(resultLocation).toHaveProperty('state')
    })
  })
})
