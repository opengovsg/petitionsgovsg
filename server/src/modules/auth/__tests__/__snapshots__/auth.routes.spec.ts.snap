// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/auth Get / retrieves authenticated user 1`] = `
Object {
  "fullname": "John Doe",
  "id": "sgid",
}
`;

exports[`/auth Get / returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in.",
}
`;

exports[`/auth Get /checkpetitionowner/:id returns OK false on user is not petition owner 1`] = `false`;

exports[`/auth Get /checkpetitionowner/:id returns OK true on user is petition owner 1`] = `true`;

exports[`/auth Get /checkpetitionowner/:id returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in.",
}
`;

exports[`/auth Get /fullname returns OK with fullname 1`] = `
Object {
  "fullname": "John Doe",
}
`;

exports[`/auth Get /fullname returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in.",
}
`;
